Current checkpoint:

```
pulp_1          | ('pulp [b8619d0338e34530881e4890bea1acf7]: ::ffff:127.0.0.1 - - [30/Jul/2024:20:00:51 +0000] "DELETE /api/galaxy/_ui/v2/role_definitions/23/ HTTP/1.0" 204 0 "-" "python-requests/2.31.0"',)
```

I see this request from the _teardown_

but the test doesn't like that

```
/tmp/gng_testing/lib64/python3.11/site-packages/requests/models.py:975: JSONDecodeError

The above exception was the direct cause of the following exception:

galaxy_client = <bound method GalaxyKitClient.gen_authorized_client of <galaxy_ng.tests.integration.utils.iqe_utils.GalaxyKitClient object at 0x7f27cb72fbd0>>

    @pytest.fixture(scope="module")
    def custom_ns_role(galaxy_client):
        gc = galaxy_client("admin")
        list_r = gc.get("/api/galaxy/_ui/v2/role_definitions/?name=My own Namespace admin role")
        if list_r["count"] == 1:
            r = list_r["results"][0]
            yield r
        elif list_r["count"] > 1:
            raise RuntimeError("Found too many role_definitions with expected name")
        else:
            r = gc.post(
                "/api/galaxy/_ui/v2/role_definitions/",
                body={
                    "name": "My own Namespace admin role",
                    "permissions": [
                        "galaxy.change_namespace",
                        "galaxy.delete_namespace",
                        "galaxy.view_namespace",
                        "galaxy.view_collectionimport",
                    ],
                    "content_type": None,
                },
            )
            yield r
        with pytest.raises(GalaxyClientError) as ctx:
>           gc.delete(f'/api/galaxy/_ui/v2/role_definitions/{r["id"]}/')

galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/tmp/gng_testing/lib64/python3.11/site-packages/galaxykit/client.py:376: in delete
    return self._http("delete", path, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <galaxykit.client.GalaxyClient object at 0x7f27cb904250>, method = 'delete'
path = '/api/galaxy/_ui/v2/role_definitions/23/', args = (), kwargs = {}
url = 'http://localhost:5001/api/galaxy/_ui/v2/role_definitions/23/'
headers = {'Accept': 'application/json', 'Authorization': 'Token abcdefghijklmnopqrstuvwxyz1234567894'}
parse_json = True, relogin = True, resp = <Response [204]>

    def _http(self, method, path, *args, **kwargs):
        url = urljoin(self.galaxy_root, path)
        headers = kwargs.pop("headers", self.headers)
        parse_json = kwargs.pop("parse_json", True)
        relogin = kwargs.pop("relogin", True)
        resp = send_request_with_retry_if_504(
            method, url, headers=headers, verify=self.https_verify, *args, **kwargs
        )
        self.response = resp
        if "Invalid JWT token" in resp.text:
            resp = self._retry_if_expired_token(method, url, headers, *args, **kwargs)
        if parse_json:
            try:
                json_data = resp.json()
            except JSONDecodeError as exc:
                logging.error(
                    f"Cannot parse expected JSON response ({url}): {resp.text}"
                )
>               raise ValueError("Failed to parse JSON response from API") from exc
E               ValueError: Failed to parse JSON response from API

/tmp/gng_testing/lib64/python3.11/site-packages/galaxykit/client.py:256: ValueError
```

I don't think I should catch ValueError, but I guess no harm in it?

--- a/galaxy_ng/app/signals/handlers.py
+++ b/galaxy_ng/app/signals/handlers.py
@@ -3,6 +3,10 @@ Signal handlers for the Galaxy application.
 Those signals are loaded by
 galaxy_ng.app.__init__:PulpGalaxyPluginAppConfig.ready() method.
 """
+
+import threading
+import contextlib
+
 from django.dispatch import receiver
 from django.db.models.signals import post_save
 from django.db.models.signals import post_delete
@@ -10,13 +14,14 @@ from pulp_ansible.app.models import (
     AnsibleDistribution,
     AnsibleRepository,
     Collection,
-    AnsibleNamespaceMetadata
+    AnsibleNamespaceMetadata,
 )
 from galaxy_ng.app.models import Namespace
 from pulpcore.plugin.models import ContentRedirectContentGuard
 
 from ansible_base.rbac.models import RoleTeamAssignment
 from ansible_base.rbac.models import RoleUserAssignment
+from ansible_base.rbac.models import RoleDefinition
 from pulpcore.plugin.util import assign_role
 from pulpcore.plugin.util import remove_role
 
@@ -84,25 +89,137 @@ def associate_namespace_metadata(sender, instance, created, **kwargs):
         _update_metadata()
 
 
+# Signals for synchronizing the pulp roles with DAB RBAC roles
+rbac_state = threading.local()
+
+rbac_state.pulp_action = False
+rbac_state.dab_action = False
+
+
+@contextlib.contextmanager
+def pulp_rbac_signals():
+    "Used while firing signals from pulp RBAC models to avoid infinite loops"
+    try:
+        prior_value = rbac_state.pulp_action
+        rbac_state.pulp_action = True
+        yield
+    finally:
+        rbac_state.pulp_action = prior_value
+
+
+@contextlib.contextmanager
+def dab_rbac_signals():
+    "Used while firing signals from DAB RBAC models to avoid infinite loops"
+    try:
+        prior_value = rbac_state.dab_action
+        rbac_state.dab_action = True
+        yield
+    finally:
+        rbac_state.dab_action = prior_value
+
+
+# Pulp Role to DAB RBAC RoleDefinition objects
+
+
+@receiver(post_save, sender=Role)
+def copy_role_to_role_definition(sender, instance, created, **kwargs):
+    """When a dab role is granted to a user, grant the equivalent pulp role."""
+    if rbac_state.pulp_action:
+        return
+    with pulp_rbac_signals():
+        rd = RoleDefinition.objects.filter(name=instance.name).first()
+        if not rd:
+            RoleDefinition.objects.create(name=instance.name)
+        # TODO: other fields? like description
+
+
+# DAB RBAC RoleDefinition objects to Pulp Role objects
+
+
+@receiver(post_save, sender=RoleDefinition)
+def copy_role_definition_to_role(sender, instance, created, **kwargs):
+    """When a dab role is granted to a user, grant the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        role = Role.objects.filter(name=instance.name).first()
+        if not role:
+            Role.objects.create(name=instance.name)
+        # TODO: other fields? like description
+
+
+@receiver(post_delete, sender=RoleDefinition)
+def delete_role_definition_to_role(sender, instance, **kwargs):
+    """When a dab role is granted to a user, grant the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        role = Role.objects.filter(name=instance.name).first()
+        if role:
+            role.delete()
+
+
+# m2m_changed
+# - adding
+# - removing
+# post_delete
+
+
+# Pulp UserRole and TeamRole to DAB RBAC assignments
+
+# TODO:
+# - UserRole deletion
+# - TeamRole creation
+# - TeamRole deletion
+
+
+@receiver(post_save, sender=UserRole)
+def copy_pulp_user_role(sender, instance, created, **kwargs):
+    """When a dab role is granted to a user, grant the equivalent pulp role."""
+    if rbac_state.pulp_action:
+        return
+    with pulp_rbac_signals():
+        rd = RoleDefinition.objects.get(name=instance.role.name)
+        if instance.content_object:
+            rd.give_permission(instance.user, instance.content_object)
+        else:
+            rd.give_global_permission(instance.user)
+
+
+# DAB RBAC assignments to pulp UserRole TeamRole
+
+
 @receiver(post_save, sender=RoleUserAssignment)
 def copy_dab_user_role(sender, instance, created, **kwargs):
-    '''When a dab role is granted to a user, grant the equivalent pulp role.'''
-    assign_role(instance.role_definition.name, instance.user)
+    """When a dab role is granted to a user, grant the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        assign_role(instance.role_definition.name, instance.user)
 
 
 @receiver(post_delete, sender=RoleUserAssignment)
 def delete_dab_user_role(sender, instance, **kwargs):
-    '''When a dab role is revoked from a user, revoke the equivalent pulp role.'''
-    remove_role(instance.role_definition.name, instance.user)
+    """When a dab role is revoked from a user, revoke the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        remove_role(instance.role_definition.name, instance.user)
 
 
 @receiver(post_save, sender=RoleTeamAssignment)
 def copy_dab_team_role(sender, instance, created, **kwargs):
-    '''When a dab role is granted to a team, grant the equivalent pulp role.'''
-    assign_role(instance.role_definition.name, instance.team.group)
+    """When a dab role is granted to a team, grant the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        assign_role(instance.role_definition.name, instance.team.group)
 
 
 @receiver(post_delete, sender=RoleTeamAssignment)
 def delete_dab_team_role(sender, instance, **kwargs):
-    '''When a dab role is revoked from a team, revoke the equivalent pulp role.'''
-    remove_role(instance.role_definition.name, instance.team.group)
+    """When a dab role is revoked from a team, revoke the equivalent pulp role."""
+    if rbac_state.dab_action:
+        return
+    with dab_rbac_signals():
+        remove_role(instance.role_definition.name, instance.team.group)

--- a/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
+++ b/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
@@ -50,6 +50,15 @@ def test_role_definition_options(galaxy_client):
         'shared.view_team',
     }
 
+    assert 'content_type' in post_data
+    field_data = post_data['content_type']
+    assert 'choices' in field_data
+    assert set(item['value'] for item in field_data['choices']) == {
+        'galaxy.collectionimport',
+        'galaxy.namespace',
+        'shared.team',
+    }
+
 
 @pytest.fixture(scope="module")
 def custom_ns_role(galaxy_client):

--- a/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
+++ b/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
@@ -3,55 +3,62 @@ import pytest
 
 # This tests the basic DAB RBAC contract using custom roles to do things
 
+
 def test_list_namespace_permissions(galaxy_client):
-    gc = galaxy_client('admin')
-    r = gc.get('/api/galaxy/_ui/v2/role_metadata/')
-    assert 'galaxy.namespace' in r['allowed_permissions']
-    allowed_perms = r['allowed_permissions']['galaxy.namespace']
+    gc = galaxy_client("admin")
+    r = gc.get("/api/galaxy/_ui/v2/role_metadata/")
+    assert "galaxy.namespace" in r["allowed_permissions"]
+    allowed_perms = r["allowed_permissions"]["galaxy.namespace"]
     assert set(allowed_perms) == {
-        'galaxy.change_namespace',
-        'galaxy.delete_namespace',
-        'galaxy.add_collectionimport',
-        'galaxy.change_collectionimport',
-        'galaxy.delete_collectionimport',
-        'galaxy.upload_to_namespace',
-        'galaxy.view_collectionimport',
-        'galaxy.view_namespace',
+        "galaxy.change_namespace",
+        "galaxy.delete_namespace",
+        "galaxy.add_collectionimport",
+        "galaxy.change_collectionimport",
+        "galaxy.delete_collectionimport",
+        "galaxy.upload_to_namespace",
+        "galaxy.view_collectionimport",
+        "galaxy.view_namespace",
     }
 
 
-@pytest.fixture(scope='module')
+@pytest.fixture(scope="module")
 def custom_ns_role(galaxy_client):
-    gc = galaxy_client('admin')
-    list_r = gc.get('/api/galaxy/_ui/v2/role_definitions/?name=My own Namespace admin role')
-    if list_r['count'] == 1:
-        return list_r['results'][0]
-    elif list_r['count'] > 1:
-        raise RuntimeError('Found too many role_definitions with expected name')
-    r = gc.post('/api/galaxy/_ui/v2/role_definitions/', body={
-        "name": "My own Namespace admin role",
-        "permissions": [
-            'galaxy.change_namespace',
-            'galaxy.delete_namespace',
-            'galaxy.view_namespace',
-            'galaxy.view_collectionimport'
-        ],
-        "content_type": None
-    })
-    return r    
+    gc = galaxy_client("admin")
+    list_r = gc.get("/api/galaxy/_ui/v2/role_definitions/?name=My own Namespace admin role")
+    if list_r["count"] == 1:
+        r = list_r["results"][0]
+        yield r
+    elif list_r["count"] > 1:
+        raise RuntimeError("Found too many role_definitions with expected name")
+    else:
+        r = gc.post(
+            "/api/galaxy/_ui/v2/role_definitions/",
+            body={
+                "name": "My own Namespace admin role",
+                "permissions": [
+                    "galaxy.change_namespace",
+                    "galaxy.delete_namespace",
+                    "galaxy.view_namespace",
+                    "galaxy.view_collectionimport",
+                ],
+                "content_type": None,
+            },
+        )
+        yield r
+    gc.delete(f'/api/galaxy/_ui/v2/role_definitions/{r["id"]}/')
 
 
 def test_create_custom_namespace_admin_role(custom_ns_role):
-    assert custom_ns_role['name'] == "My own Namespace admin role"
+    assert custom_ns_role["name"] == "My own Namespace admin role"
 
 
 def test_give_custom_role(galaxy_client, custom_ns_role):
-    gc = galaxy_client('admin')
-    user_r = gc.get('/api/galaxy/_ui/v2/users/')
-    assert user_r['count'] > 0
-    user = user_r['results'][0]
-    assignment = gc.post('/api/galaxy/_ui/v2/role_user_assignments/', body={
-        'role_definition': custom_ns_role['id'],
-        'user': user['id']
-    })
+    gc = galaxy_client("admin")
+    user_r = gc.get("/api/galaxy/_ui/v2/users/")
+    assert user_r["count"] > 0
+    user = user_r["results"][0]
+    assignment = gc.post(
+        "/api/galaxy/_ui/v2/role_user_assignments/",
+        body={"role_definition": custom_ns_role["id"], "user": user["id"]},
+    )
     # TODO: make a request as the user and see that it works

--- a/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
+++ b/galaxy_ng/tests/integration/dab/test_dab_rbac_contract.py
@@ -1,0 +1,57 @@
+import pytest
+
+
+# This tests the basic DAB RBAC contract using custom roles to do things
+
+def test_list_namespace_permissions(galaxy_client):
+    gc = galaxy_client('admin')
+    r = gc.get('/api/galaxy/_ui/v2/role_metadata/')
+    assert 'galaxy.namespace' in r['allowed_permissions']
+    allowed_perms = r['allowed_permissions']['galaxy.namespace']
+    assert set(allowed_perms) == {
+        'galaxy.change_namespace',
+        'galaxy.delete_namespace',
+        'galaxy.add_collectionimport',
+        'galaxy.change_collectionimport',
+        'galaxy.delete_collectionimport',
+        'galaxy.upload_to_namespace',
+        'galaxy.view_collectionimport',
+        'galaxy.view_namespace',
+    }
+
+
+@pytest.fixture(scope='module')
+def custom_ns_role(galaxy_client):
+    gc = galaxy_client('admin')
+    list_r = gc.get('/api/galaxy/_ui/v2/role_definitions/?name=My own Namespace admin role')
+    if list_r['count'] == 1:
+        return list_r['results'][0]
+    elif list_r['count'] > 1:
+        raise RuntimeError('Found too many role_definitions with expected name')
+    r = gc.post('/api/galaxy/_ui/v2/role_definitions/', body={
+        "name": "My own Namespace admin role",
+        "permissions": [
+            'galaxy.change_namespace',
+            'galaxy.delete_namespace',
+            'galaxy.view_namespace',
+            'galaxy.view_collectionimport'
+        ],
+        "content_type": None
+    })
+    return r    
+
+
+def test_create_custom_namespace_admin_role(custom_ns_role):
+    assert custom_ns_role['name'] == "My own Namespace admin role"
+
+
+def test_give_custom_role(galaxy_client, custom_ns_role):
+    gc = galaxy_client('admin')
+    user_r = gc.get('/api/galaxy/_ui/v2/users/')
+    assert user_r['count'] > 0
+    user = user_r['results'][0]
+    assignment = gc.post('/api/galaxy/_ui/v2/role_user_assignments/', body={
+        'role_definition': custom_ns_role['id'],
+        'user': user['id']
+    })
+    # TODO: make a request as the user and see that it works

